site-permissions/site-permissions-impl/src/main/java/com/duckduckgo/site/permissions/impl/SitePermissionsDialogActivityLauncher.kt

```diff
@@ -0,0 +1,206 @@
+        context: Context,
+        url: String,
+        permissionsRequested: Array<String>,
+        request: PermissionRequest
+    ) {
+        sitePermissionRequest = request
+        this.context = context
+
+        when {
+            permissionsRequested.size == 2 -> {
+                showSitePermissionsRationaleDialog(R.string.sitePermissionsMicAndCameraDialogTitle, url, this::askForMicAndCameraPermissions)
+            }
+            permissionsRequested.contains(PermissionRequest.RESOURCE_AUDIO_CAPTURE) -> {
+                showSitePermissionsRationaleDialog(R.string.sitePermissionsMicDialogTitle, url, this::askForMicPermissions)
+            }
+            permissionsRequested.contains(PermissionRequest.RESOURCE_VIDEO_CAPTURE) -> {
+                showSitePermissionsRationaleDialog(R.string.sitePermissionsCameraDialogTitle, url, this::askForCameraPermissions)
+            }
+        }
+    }
+
+    private fun showSitePermissionsRationaleDialog(
+        @StringRes titleRes: Int,
+        url: String,
+        onPermissionAllowed: () -> Unit
+    ) {
+        val dialog = AlertDialog.Builder(context)
+        val view: View = LayoutInflater.from(context).inflate(layout.dialog_site_permissions, null)
+        val title = view.findViewById<TextView>(R.id.sitePermissionsDialogTitle)
+        title.text = String.format(context.getString(titleRes), url.websiteFromGeoLocationsApiOrigin())
+        dialog.setView(view)
+        dialog.apply {
+            setPositiveButton(R.string.sitePermissionsDialogAllowButton) { dialog, _ ->
+                onPermissionAllowed()
+            }
+            setNegativeButton(R.string.sitePermissionsDialogDenyButton) { dialog, _ ->
+                Toast.makeText(context, "Deny", Toast.LENGTH_SHORT).show()
+            }
+            show()
+        }
+    }
+
+    private fun askForMicAndCameraPermissions() {
+        permissionRequested = SitePermissionsRequestedType.CAMERA_AND_AUDIO
+        when {
+            systemPermissionsHelper.hasMicPermissionsGranted() && systemPermissionsHelper.hasCameraPermissionsGranted() -> {
+                systemPermissionGranted(SitePermissionsRequestedType.CAMERA_AND_AUDIO)
+            }
+            systemPermissionsHelper.hasMicPermissionsGranted() -> {
+                systemPermissionsHelper.requestPermission(Manifest.permission.CAMERA)
+            }
+            systemPermissionsHelper.hasCameraPermissionsGranted() -> {
+                systemPermissionsHelper.requestMultiplePermissions(
+                    arrayOf(
+                        Manifest.permission.RECORD_AUDIO,
+                        Manifest.permission.MODIFY_AUDIO_SETTINGS
+                    )
+                )
+            }
+            else -> {
+                systemPermissionsHelper.requestMultiplePermissions(
+                    arrayOf(
+                        Manifest.permission.RECORD_AUDIO,
+                        Manifest.permission.MODIFY_AUDIO_SETTINGS,
```

Comment:
Just out of curiosity, do we need `MODIFY_AUDIO_SETTINGS` and `RECORD_AUDIO` to use the microphone?

Reply:
yes, in the beginning it wasn't working because I was only requesting `RECORD_AUDIO. Therefore, it took me a while to realised it was because we need to grant access to both of them for websites to work 😅